I did stuff

ZachPlunkett opened the pull request:
Applied a markdown linter to README.md. Fixed return issues with recipes.

Added a hello world example based upon the readme.

ZachPlunkett committed:
<examples>: Added basic examples

Added an example for Execute and ExecuteInto. This should give a
perspective user a quick way to understand library usage.

fixes: #12

author: ZachPlunkett <Zach13plunkett@gmail.com>

ZachPlunkett committed:
linted README.md and corrected recipes

ZachPlunkett committed:
Added a hello world type example

My smol brain needed this to understand what was going on. Maybe someone else will find this useful.

ZachPlunkett committed:
Fixed issues with linting (\*)

ShaileshSurya commented on example/main.go:
```diff
@@ -0,0 +1,64 @@
+package main
```
@ZachPlunkett  Thank you so much for example/main.go. This was needed. Great work with the ideation.

ZachPlunkett replied:
Thank you for the feedback!

ShaileshSurya reviewed (changes requested):
Blocking the PR till, We fully review this.

ShaileshSurya commented on Readme.md:
```diff
@@ -1,54 +1,54 @@
 
-map := make(map[string]interface{})
-resp,err:= client.ExecuteInto(request, &map)
+m := make(map[string]interface{})
```
Can we rename the `m` to map or more descriptive name like, responseMap?

ZachPlunkett replied:
Fixed

ShaileshSurya replied:
I think we missed this. I think we should `m` to `responseMap`, We want to make our intention clear.

ZachPlunkett replied:
Fixed for real this time

ShaileshSurya commented on Readme.md:
```diff
@@ -1,54 +1,54 @@
-or 
+// or
 
 model := Employee{}
```
If we choose to change the `m` to `responseMap`, just to stay consistent we should rename this to `responseModel`.

ZachPlunkett replied:
Fixed

ShaileshSurya commented on Readme.md:
```diff
@@ -74,56 +76,55 @@ WithContext(ctx context.Context)
-		WithHeaders("user-id", "10062").
-		WithRequestParams("department", "HR").
-		Build()
+Get().
```
We removed the indentation it seems.

ZachPlunkett replied:
Fixed

ShaileshSurya commented on Readme.md:
```diff
@@ -74,56 +76,55 @@ WithContext(ctx context.Context)
-		WithHeaders("user-id", "10062").
-		WithRequestParams("department", "HR").
-		Build()
+Get().
```
Indentation can be improved here as well.

ZachPlunkett replied:
Fixed

ShaileshSurya commented on Readme.md:
```diff
@@ -74,56 +76,55 @@ WithContext(ctx context.Context)
-        	WithRequestParams("post","manager").
-        	WithRequestParams("centre","pune")
-		Build()
+WithTemplate(reqTemp).
```
Indentation

ZachPlunkett replied:
Fixed

ShaileshSurya commented on example/main.go:
```diff
@@ -0,0 +1,64 @@
+	"github.com/ShaileshSurya/hammer"
+)
+
+func handleErr(err error) {
```
`handleErr` is in widespread use. But if personally prefer verbosity to than using this function. It will help us with the code readability. 

We can remove this function altogether.

ZachPlunkett replied:
Fixed

ShaileshSurya commented on example/main.go:
```diff
@@ -0,0 +1,64 @@
+
+	fmt.Printf("%+v\n\n", hw)
+
+	// Execute the request and feed response body into an empty interface
```
`into an empty interface`, We meant `map` here right?

ZachPlunkett replied:
Fixed

ShaileshSurya commented on example/main.go:
```diff
@@ -0,0 +1,64 @@
+	}
+}
+
+func main() {
```
We can break the main into two different functions 
1. `ExampleForExecute()`
2. `ExampleForExecuteInto()`

As this is will be more visible to the library reader. We want to make our intent as much clear as possible. 

Or This can be broken into two different files altogether. One for `execute` and another for `executeInto`.

ZachPlunkett replied:
Changed into two separate files.

examples/execute/main.go
example/execute_into/main.go

ShaileshSurya commented on example/main.go:
```diff
@@ -0,0 +1,64 @@
+
+	fmt.Printf("%+v\n\n", m)
+
+	// Execute the request and feed response body into a struct
```
We might not need this comment. As the `ExecuteInto` function name itself is self explanatory.

ZachPlunkett replied:
Comment was changed to 
// ExecuteInto a struct 
mainly to separate the two examples using ExecuteInto

ShaileshSurya commented on example/main.go:
```diff
@@ -0,0 +1,64 @@
+	fmt.Printf("%+v\n\n", hw)
+
+	// Execute the request and feed response body into an empty interface
+	m := make(map[string]interface{})
```
Can we rename the `m` to more descriptive. Like `responseMap`

ZachPlunkett replied:
Fixed

ZachPlunkett committed:
PR changes implemented

Increaded verbosity/readablility with
    m -> responseMap
    handleErr -> if err != nil {...

Fixed indentation in README.md examples

Broke the single example into 2 sepearte examples

ZachPlunkett commented:
I implemented the changes you suggested. Thanks for the feedback.

ShaileshSurya commented:
@ZachPlunkett  Thats a real good progress. Can you just reply to the review comments? Like saying Fixed or raising any reservation that you may have. It will help speed up the review process.

ShaileshSurya commented on examples/execute/main.go:
```diff
@@ -1,64 +1,51 @@
-
-	fmt.Printf("%+v\n\n", hw2)
-}
+package main
```
The changes looks good to me. We might want to use the file path structure such as `examples\execute.go`

The user might run this with `go run execute.go`

With this file structure we wont need to create separate folder for each of the API. In future iteration we might want to add examples for RequestBuilder or in that case ResponseBuilder as well.

ZachPlunkett replied:
Fixed

ShaileshSurya replied:
@ZachPlunkett Looks like this change broke the CI pipeline. You were right, the proposed file structure by you was right. My bad. 

![image](https://user-images.githubusercontent.com/22809827/96751558-f9411880-13ea-11eb-948d-e803d5b62b0c.png)

ZachPlunkett replied:
Fixed by reverting back to original structure

ShaileshSurya commented on Readme.md:
```diff
@@ -26,7 +26,7 @@ err:= client.ExecuteInto(request, &m)
 
 // or
```
Can we add comments like `Fixed` or `done` or raise any concern that we may have with the review comments. It will help speed up the review process?

ZachPlunkett replied:
Added the Fixed comments and I also resolved them. Let me know what if that isn't what I was supposed to do.

ZachPlunkett committed:
readable example path/file names

ZachPlunkett committed:
Merge remote-tracking branch 'upstream/master' into master

ShaileshSurya commented on examples/execute_into/main.go:
```diff
@@ -0,0 +1,48 @@
+)
+
+func main() {
+	// init client
```
We dont need this comment.

ShaileshSurya replied:
We should remove from execute.go

ZachPlunkett replied:
Fixed

ShaileshSurya commented on examples/execute_into/main.go:
```diff
@@ -0,0 +1,48 @@
+	client := hammer.New()
+
+	// build request
+	request, err := hammer.RequestBuilder().
```
with `hammer.RequestBuilder` it is given that here we are trying to build request. We do not need this comment. 

I prefer to use comments very sparingly.

ShaileshSurya replied:
We should remove this from execute.go

ZachPlunkett replied:
Fixed

ShaileshSurya commented on examples/execute_into/main.go:
```diff
@@ -0,0 +1,48 @@
+		Place    string
+	}
+
+	var hw HelloWorld
```
Can we add the comment from line 35 Here and remove it from the line 35.

ZachPlunkett replied:
Fixed

ShaileshSurya commented:
@ZachPlunkett the PR is almost shaped up. Just Few review comment and we are good to merge. Make sure to squash and merge the commits into the single commit.

ShaileshSurya commented on examples/execute_into.go:
```diff
@@ -0,0 +1,48 @@
+
+func main() {
+	// init client
+	client := hammer.New()
```
We should Remove this comment. We dont need this.

ZachPlunkett replied:
Fixed

ShaileshSurya commented on examples/execute_into.go:
```diff
@@ -0,0 +1,48 @@
+	// init client
+	client := hammer.New()
+
+	// build request
```
We don't need this comment. We should remove this.

ZachPlunkett replied:
Fixed

ShaileshSurya commented on examples/execute_into.go:
```diff
@@ -0,0 +1,48 @@
+	}
+
+	var hw HelloWorld
+	err = client.ExecuteInto(request, &hw)
```
Add the comment on line 35 to here. Remove it from the line 35.

ZachPlunkett replied:
Fixed

ShaileshSurya commented on examples/execute.go:
```diff
@@ -0,0 +1,51 @@
+
+func main() {
+	// init client
+	client := hammer.New()
```
We dont need this comment.

ZachPlunkett replied:
Fixed

ShaileshSurya commented on examples/execute.go:
```diff
@@ -0,0 +1,51 @@
+	client := hammer.New()
+
+	// build request
+	request, err := hammer.RequestBuilder().
```
We dont need this comment.

ZachPlunkett replied:
Fixed

ShaileshSurya commented on Readme.md:
```diff
@@ -1,54 +1,54 @@
-
```
There is something real bad happening with my git. Execuse me if I give review on stale commits. Or repeat my comments.

ZachPlunkett replied:
All good, I am learning a lot doing this so I don't know what you would need to excuse yourself for anyways.

ZachPlunkett committed:
Merge branch 'master' of https://github.com/ZachPlunkett/hammer into master

ZachPlunkett committed:
<examples> Fixed example path/names structure

The travis-ci broke with multiple mains in 1 folder. So reverted back to
structure of examples/execute/main.go and examples/execute_into/main.go

Fixes: #12

author: ZachPlunkett <Zach13plunkett@gmail.com>

ZachPlunkett commented:
I tried my best to rebase and squash my commits. But I couldn't get it into 1. When I did it then told me I was behind master and I needed to do another git pull and then there were conflicts all over the place. 

I am sure I could have done it better but that is the best I have got so far.

ZachPlunkett commented:
> I tried my best to rebase and squash my commits. But I couldn't get it into 1. When I did it then told me I was behind master and I needed to do another git pull and then there were conflicts all over the place.
> 
> I am sure I could have done it better but that is the best I have got so far.

Well that was a lie. I guess I didn't squash any of the commits like I thought I did. I was trying `git rebase -i <your last commit>` and then I marked my first commit for rename and all the rest as squash. I am doing something wrong obviously just I don't intimately enough to do this. Are you able to squash and merge the PR? 

Sorry for the complication.

ZachPlunkett committed:
<doc>: corrected m to responseMap

ShaileshSurya commented on Readme.md:
```diff
@@ -21,7 +21,7 @@ resp, err:= client.Execute(request)
 
-m := make(map[string]interface{})
+responseMap := make(map[string]interface{})
 err:= client.ExecuteInto(request, &m)
```
`err:= client.ExecuteInto(request, &responseMap)`, This also needs to be refactored.

ShaileshSurya commented:
Closing this PR in favour of #14. Thank you @ZachPlunkett  for your contribution.
